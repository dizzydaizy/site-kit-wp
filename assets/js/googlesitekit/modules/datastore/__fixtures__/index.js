/**
 * Modules datastore fixtures.
 *
 * Site Kit by Google, Copyright 2021 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import modules from './list.json';

// Only Search Console and Site Verification are always active.
const alwaysActive = [ 'search-console', 'site-verification' ];

/**
 * Makes a copy of the modules with the given module activation set.
 *
 * @since 1.18.0
 *
 * @param {...string} slugs Active module slugs.
 * @return {Object[]} Array of module objects.
 */
export function withActive( ...slugs ) {
	const activeSlugs = alwaysActive.concat( slugs );
	return modules.map( ( module ) => ( {
		...module,
		active: activeSlugs.includes( module.slug ),
	} ) );
}

/**
 * Makes a copy of the modules with the given module connection set.
 *
 * @since n.e.x.t
 *
 * @param {...string} slugs Active module slugs.
 * @return {Object[]} Array of module objects.
 */
export function withConnected( ...slugs ) {
	const activeSlugs = alwaysActive.concat( slugs );
	return withActive( ...slugs ).map( ( module ) => ( {
		...module,
		connected: activeSlugs.includes( module.slug ),
	} ) );
}

export default withActive();
