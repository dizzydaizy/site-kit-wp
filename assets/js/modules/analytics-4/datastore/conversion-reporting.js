/**
 * `modules/analytics-4` data store: conversion reporting.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import invariant from 'invariant';

/**
 * Internal dependencies
 */
import API from 'googlesitekit-api';
import {
	commonActions,
	combineStores,
	createRegistrySelector,
	createReducer,
} from 'googlesitekit-data';
import { MODULES_ANALYTICS_4 } from './constants';
import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
import { negateDefined } from '../../../util/negate';
import { CORE_USER } from '../../../googlesitekit/datastore/user/constants';
import { getKeyMetricsConversionEventWidgets } from '../../../googlesitekit/datastore/util/conversion-reporting';

function hasConversionReportingEventsOfType( propName ) {
	return createRegistrySelector( ( select ) => () => {
		const inlineData =
			select(
				MODULES_ANALYTICS_4
			).getConversionReportingEventsChange() || {};

		// Here we double-negate the value in order to cast it to a boolean, but only if it's not undefined.
		return negateDefined( negateDefined( inlineData[ propName ]?.length ) );
	} );
}

const dismissNewConversionReportingEventsStore = createFetchStore( {
	baseName: 'dismissNewConversionReportingEvents',
	controlCallback: () => {
		return API.set(
			'modules',
			'analytics-4',
			'clear-conversion-reporting-new-events'
		);
	},
	reducerCallback: ( state, values ) => {
		if ( values === false ) {
			return state;
		}

		return {
			...state,
			detectedEventsChange: {
				...state.detectedEventsChange,
				newEvents: [],
			},
		};
	},
} );

const dismissLostConversionReportingEventsStore = createFetchStore( {
	baseName: 'dismissLostConversionReportingEvents',
	controlCallback: () => {
		return API.set(
			'modules',
			'analytics-4',
			'clear-conversion-reporting-lost-events'
		);
	},
	reducerCallback: ( state, values ) => {
		if ( values === false ) {
			return state;
		}

		return {
			...state,
			detectedEventsChange: {
				...state.detectedEventsChange,
				lostEvents: [],
			},
		};
	},
} );

// Actions.
const RECEIVE_CONVERSION_REPORTING_INLINE_DATA =
	'RECEIVE_CONVERSION_REPORTING_INLINE_DATA';

export const initialState = {
	detectedEventsChange: undefined,
};

export const resolvers = {
	*getConversionReportingEventsChange() {
		const registry = yield commonActions.getRegistry();

		if (
			registry
				.select( MODULES_ANALYTICS_4 )
				.getConversionReportingEventsChange()
		) {
			return;
		}

		if ( ! global._googlesitekitModulesData ) {
			global.console.error( 'Could not load modules data.' );
			return;
		}

		const { newEvents, lostEvents } =
			global._googlesitekitModulesData[ 'analytics-4' ];

		yield actions.receiveConversionReportingInlineData( {
			newEvents,
			lostEvents,
		} );
	},
};

export const actions = {
	/**
	 * Dismiss new conversion reporting events.
	 *
	 * @since 1.138.0
	 *
	 * @return {boolean} Transient deletion response.
	 */
	dismissNewConversionReportingEvents() {
		return dismissNewConversionReportingEventsStore.actions.fetchDismissNewConversionReportingEvents();
	},

	/**
	 * Dismiss lost conversion reporting events.
	 *
	 * @since 1.138.0
	 *
	 * @return {boolean} Transient deletion response.
	 */
	dismissLostConversionReportingEvents() {
		return dismissLostConversionReportingEventsStore.actions.fetchDismissLostConversionReportingEvents();
	},

	/**
	 * Stores conversion reporting inline data in the datastore.
	 *
	 * @since 1.140.0
	 * @private
	 *
	 * @param {Object} data Inline data, usually supplied via a global variable from PHP.
	 * @return {Object} Redux-style action.
	 */
	receiveConversionReportingInlineData( data ) {
		invariant( data, 'data is required.' );

		return {
			payload: { data },
			type: RECEIVE_CONVERSION_REPORTING_INLINE_DATA,
		};
	},
};

export const reducer = createReducer( ( state, { payload, type } ) => {
	switch ( type ) {
		case RECEIVE_CONVERSION_REPORTING_INLINE_DATA: {
			const { newEvents, lostEvents } = payload.data;

			state.detectedEventsChange = { newEvents, lostEvents };
			break;
		}

		default: {
			break;
		}
	}
} );

export const selectors = {
	/**
	 * Checks whether the provided conversion reporting events are available.
	 *
	 * @since 1.135.0
	 *
	 * @param {Object}               state  Data store's state.
	 * @param {string|Array<string>} events Conversion reporting events to check.
	 * @return {(boolean|undefined)} True if all provided custom dimensions are available, otherwise false. Undefined if available custom dimensions are not loaded yet.
	 */
	hasConversionReportingEvents: createRegistrySelector(
		( select ) => ( state, events ) => {
			// Ensure events is always an array, even if a string is passed.
			const eventsToCheck = Array.isArray( events ) ? events : [ events ];

			const detectedEvents =
				select( MODULES_ANALYTICS_4 ).getDetectedEvents();

			if ( ! detectedEvents?.length ) {
				return false;
			}

			return eventsToCheck.some( ( event ) =>
				detectedEvents.includes( event )
			);
		}
	),

	/**
	 * Gets all conversion reporting inline data from this data store.
	 *
	 * Not intended to be used publicly; this is largely here so other selectors can
	 * request data using the selector/resolver pattern.
	 *
	 * @since 1.140.0
	 * @private
	 *
	 * @param {Object} state Data store's state.
	 * @return {(Object|undefined)} Conversion reporting inline data.
	 */
	getConversionReportingEventsChange( state ) {
		return state.detectedEventsChange;
	},

	/**
	 * Checks if newEvents are present.
	 *
	 * @since 1.140.0
	 *
	 * @param {Object} state Data store's state.
	 * @return {boolean|undefined} TRUE if `newEvents` are present, FALSE otherwise.
	 */
	hasNewConversionReportingEvents:
		hasConversionReportingEventsOfType( 'newEvents' ),

	/**
	 * Checks if lostEvents are present.
	 *
	 * @since 1.140.0
	 *
	 * @param {Object} state Data store's state.
	 * @return {boolean|undefined} TRUE if `lostEvents` are present, FALSE otherwise.
	 */
	hasLostConversionReportingEvents:
		hasConversionReportingEventsOfType( 'lostEvents' ),

	/**
	 * Checks if there are key metrics widgets connected with the detected events for the supplied purpose answer.
	 *
	 * @since n.e.x.t
	 *
	 * @param {string}  purpose      Value of saved site purpose from user input settings.
	 * @param {boolean} useNewEvents Flag inclusion of detected new events, otherwise initial detected events will be used.
	 * @return {boolean} TRUE if current site purpose will have any ACR key metrics widgets assigned to it, FALSE otherwise.
	 */
	haveConversionEventsForTailoredMetrics: createRegistrySelector(
		( select ) => ( state, purpose, useNewEvents ) => {
			const userInputSettings =
				select( CORE_USER ).getUserInputSettings();
			const purposeAnswerValue =
				purpose ?? userInputSettings?.purpose?.values?.[ 0 ];

			const purposeTailoredMetrics = select(
				CORE_USER
			).getAnswerBasedMetrics( purposeAnswerValue, true );

			const conversionEventWidgets =
				getKeyMetricsConversionEventWidgets();

			const conversionReportingEventsChange = useNewEvents
				? select(
						MODULES_ANALYTICS_4
				  ).getConversionReportingEventsChange()?.newEvents
				: select( MODULES_ANALYTICS_4 ).getDetectedEvents();

			return conversionReportingEventsChange?.reduce( ( acc, event ) => {
				// If a match has already been found, no need to continue.
				if ( acc ) {
					return true;
				}

				// Check if any item in conversionEventWidgets exists in purposeTailoredMetrics.
				return conversionEventWidgets[ event ].some( ( widget ) =>
					purposeTailoredMetrics.includes( widget )
				);
			}, false );
		}
	),
};

export default combineStores(
	dismissNewConversionReportingEventsStore,
	dismissLostConversionReportingEventsStore,
	{
		initialState,
		actions,
		resolvers,
		selectors,
		reducer,
	}
);
