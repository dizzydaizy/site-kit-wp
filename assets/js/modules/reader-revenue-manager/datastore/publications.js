/**
 * `modules/reader-revenue-manager` data store: publications.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import invariant from 'invariant';
import { isPlainObject } from 'lodash';

/**
 * Internal dependencies.
 */
import API from 'googlesitekit-api';
import { CORE_UI } from '../../../googlesitekit/datastore/ui/constants';
import { commonActions, combineStores } from 'googlesitekit-data';
import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
import { createValidatedAction } from '../../../googlesitekit/data/utils';
import {
	MODULES_READER_REVENUE_MANAGER,
	READER_REVENUE_MANAGER_MODULE_SLUG,
	PUBLICATION_ONBOARDING_STATES,
	UI_KEY_READER_REVENUE_MANAGER_SHOW_PUBLICATION_APPROVED_NOTIFICATION,
} from './constants';
import { actions as errorStoreActions } from '../../../googlesitekit/data/create-error-store';

const fetchGetPublicationsStore = createFetchStore( {
	baseName: 'getPublications',
	controlCallback: () =>
		API.get(
			'modules',
			READER_REVENUE_MANAGER_MODULE_SLUG,
			'publications',
			{},
			{ useCache: false }
		),
	reducerCallback: ( state, publications ) => ( { ...state, publications } ),
} );

const baseInitialState = {
	publications: undefined,
};

const baseActions = {
	/**
	 * Synchronizes the onboarding state of the publication with the API.
	 * Updates the settings on the server.
	 *
	 * @since 1.132.0
	 *
	 * @return {void}
	 */
	*syncPublicationOnboardingState() {
		const registry = yield commonActions.getRegistry();

		const settings = yield commonActions.await(
			registry
				.resolveSelect( MODULES_READER_REVENUE_MANAGER )
				.getSettings()
		);

		const publicationID = registry
			.select( MODULES_READER_REVENUE_MANAGER )
			.getPublicationID();

		// If there is no publication ID in state, do not attempt to sync
		// the onboarding state.
		if ( publicationID === undefined ) {
			return;
		}

		yield baseActions.resetPublications();

		const publications = yield commonActions.await(
			registry
				.resolveSelect( MODULES_READER_REVENUE_MANAGER )
				.getPublications()
		) || [];

		const publication = publications.find(
			// eslint-disable-next-line sitekit/acronym-case
			( { publicationId } ) => publicationId === publicationID
		);

		// If the publication is not found, do not attempt to sync the onboarding state.
		if ( ! publication ) {
			return;
		}

		const { onboardingState: newOnboardingState } = publication;

		registry
			.dispatch( MODULES_READER_REVENUE_MANAGER )
			.setPublicationOnboardingState( newOnboardingState );

		const hasPublicationIDChanged = registry
			.select( MODULES_READER_REVENUE_MANAGER )
			.hasSettingChanged( 'publicationID' );

		// Only save the synced onboarding state if there is a saved publication ID.
		if ( hasPublicationIDChanged ) {
			return;
		}

		yield commonActions.await(
			registry.dispatch( MODULES_READER_REVENUE_MANAGER ).saveSettings()
		);

		const { publicationOnboardingState: currentOnboardingState } = settings;

		// If the onboarding state changes to complete from a non-empty state, set
		// the key in CORE_UI to trigger the notification.
		if (
			!! currentOnboardingState &&
			newOnboardingState !== currentOnboardingState &&
			newOnboardingState ===
				PUBLICATION_ONBOARDING_STATES.ONBOARDING_COMPLETE
		) {
			registry
				.dispatch( CORE_UI )
				.setValue(
					UI_KEY_READER_REVENUE_MANAGER_SHOW_PUBLICATION_APPROVED_NOTIFICATION,
					true
				);
		}
	},

	/**
	 * Finds a matched publication.
	 *
	 * @since 1.132.0
	 *
	 * @return {Object|null} Matched publication; `null` if none found.
	 */
	*findMatchedPublication() {
		const { resolveSelect } = yield commonActions.getRegistry();
		const publications = yield commonActions.await(
			resolveSelect( MODULES_READER_REVENUE_MANAGER ).getPublications()
		);

		if ( publications.length === 0 ) {
			return null;
		}

		if ( publications.length === 1 ) {
			return publications[ 0 ];
		}

		const completedOnboardingPublication = publications.find(
			( publication ) =>
				publication.onboardingState ===
				PUBLICATION_ONBOARDING_STATES.ONBOARDING_COMPLETE
		);

		return completedOnboardingPublication || publications[ 0 ];
	},

	/**
	 * Resets the publications data in the store.
	 *
	 * @since 1.133.0
	 *
	 * @return {Object} The dispatched action results.
	 */
	*resetPublications() {
		const registry = yield commonActions.getRegistry();

		yield {
			type: 'RESET_PUBLICATIONS',
		};

		yield errorStoreActions.clearErrors( 'getPublications' );

		return registry
			.dispatch( MODULES_READER_REVENUE_MANAGER )
			.invalidateResolutionForStoreSelector( 'getPublications' );
	},

	/**
	 * Sets the given publication in the store.
	 *
	 * @since 1.133.0
	 *
	 * @param {Object} publication The publiation object.
	 * @return {Object} A Generator function.
	 */
	selectPublication: createValidatedAction(
		( publication ) => {
			invariant(
				isPlainObject( publication ),
				'A valid publication object is required.'
			);

			[ 'publicationId', 'onboardingState' ].forEach( ( key ) => {
				invariant(
					publication.hasOwnProperty( key ),
					`The publication object must contain ${ key }`
				);
			} );
		},
		// `publicationId` is the identifier used by the API.
		// eslint-disable-next-line sitekit/acronym-case
		function* ( { publicationId: publicationID, onboardingState } ) {
			const registry = yield commonActions.getRegistry();

			return registry
				.dispatch( MODULES_READER_REVENUE_MANAGER )
				.setSettings( {
					publicationID,
					publicationOnboardingState: onboardingState,
				} );
		}
	),
};

const baseControls = {};

const baseReducer = ( state, { type } ) => {
	switch ( type ) {
		case 'RESET_PUBLICATIONS':
			return {
				...state,
				publications: baseInitialState.publications,
			};
		default:
			return state;
	}
};

const baseResolvers = {
	*getPublications() {
		const registry = yield commonActions.getRegistry();
		// Only fetch publications if there are none in the store.
		const publications = registry
			.select( MODULES_READER_REVENUE_MANAGER )
			.getPublications();
		if ( publications === undefined ) {
			yield fetchGetPublicationsStore.actions.fetchGetPublications();
		}
	},
};

const baseSelectors = {
	/**
	 * Gets list of publications associated with the account.
	 *
	 * @since 1.132.0
	 *
	 * @param {Object} state Data store's state.
	 * @return {(Array.<Object>|undefined)} An array of publications; `undefined` if not loaded.
	 */
	getPublications( state ) {
		return state.publications;
	},
};

const store = combineStores( fetchGetPublicationsStore, {
	initialState: baseInitialState,
	actions: baseActions,
	controls: baseControls,
	reducer: baseReducer,
	resolvers: baseResolvers,
	selectors: baseSelectors,
} );

export const initialState = store.initialState;
export const actions = store.actions;
export const controls = store.controls;
export const reducer = store.reducer;
export const resolvers = store.resolvers;
export const selectors = store.selectors;

export default store;
