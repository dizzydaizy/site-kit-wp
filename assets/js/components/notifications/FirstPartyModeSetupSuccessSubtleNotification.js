/**
 * FirstPartyModeSetupSuccessSubtleNotification component.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

/**
 * WordPress dependencies
 */
import { __ } from '@wordpress/i18n';

/**
 * Internal dependencies
 */
import { CORE_MODULES } from '../../googlesitekit/modules/datastore/constants';
import SubtleNotification from '../../googlesitekit/notifications/components/layout/SubtleNotification';
import Dismiss from '../../googlesitekit/notifications/components/common/Dismiss';
import { useSelect } from 'googlesitekit-data';

export const FIRST_PARTY_MODE_SETUP_SUCCESS_NOTIFICATION =
	'setup-success-notification-fpm';

export default function FirstPartyModeSetupSuccessSubtleNotification( {
	id,
	Notification,
} ) {
	const isAnalyticsConnected = useSelect( ( select ) =>
		select( CORE_MODULES ).isModuleConnected( 'analytics-4' )
	);

	const isAdsConnected = useSelect( ( select ) =>
		select( CORE_MODULES ).isModuleConnected( 'ads' )
	);

	function getDescription() {
		if ( isAnalyticsConnected && isAdsConnected ) {
			return __(
				'You can always disable it in Analytics or Ads settings',
				'google-site-kit'
			);
		}

		if ( isAnalyticsConnected ) {
			return __(
				'You can always disable it in Analytics settings',
				'google-site-kit'
			);
		}

		// If Analytics is not connected, Ads will be.
		return __(
			'You can always disable it in Ads settings',
			'google-site-kit'
		);
	}

	return (
		<Notification>
			<SubtleNotification
				title={ __(
					'You successfully enabled First-party mode!',
					'google-site-kit'
				) }
				description={ getDescription() }
				dismissCTA={
					<Dismiss
						id={ id }
						primary={ false }
						dismissLabel={ __( 'Got it', 'google-site-kit' ) }
					/>
				}
			/>
		</Notification>
	);
}

FirstPartyModeSetupSuccessSubtleNotification.propTypes = {
	id: PropTypes.string.isRequired,
	Notification: PropTypes.elementType.isRequired,
};
