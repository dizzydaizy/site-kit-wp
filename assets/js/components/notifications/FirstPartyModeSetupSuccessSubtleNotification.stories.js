/**
 * AudienceSegmentationSetupSuccessSubtleNotification Component Stories.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import { provideModules } from '../../../../tests/js/utils';
import WithRegistrySetup from '../../../../tests/js/WithRegistrySetup';
import { CORE_SITE } from '../../googlesitekit/datastore/site/constants';
import { withNotificationComponentProps } from '../../googlesitekit/notifications/util/component-props';
import FirstPartyModeSetupSuccessSubtleNotification from './FirstPartyModeSetupSuccessSubtleNotification';

const NotificationWithComponentProps = withNotificationComponentProps(
	'setup-success-notification-fpm'
)( FirstPartyModeSetupSuccessSubtleNotification );

function Template() {
	return <NotificationWithComponentProps />;
}

export const WithAdsAndAnalytics = Template.bind( {} );
WithAdsAndAnalytics.storyName = 'With Ads and Analytics';
WithAdsAndAnalytics.args = {
	modules: [ 'ads', 'analytics-4' ],
};
WithAdsAndAnalytics.scenario = {};

export const WithAds = Template.bind( {} );
WithAds.storyName = 'With Ads';
WithAds.args = {
	modules: [ 'ads' ],
};

export const WithAnalytics = Template.bind( {} );
WithAnalytics.storyName = 'With Analytics';
WithAnalytics.args = {
	modules: [ 'analytics-4' ],
};

export default {
	title: 'Modules/FirstPartyMode/Dashboard/FirstPartyModeSetupSuccessSubtleNotification',
	decorators: [
		( Story, { args } ) => {
			const setupRegistry = ( registry ) => {
				provideModules(
					registry,
					args.modules.map( ( slug ) => ( {
						slug,
						active: true,
						connected: true,
					} ) )
				);

				registry
					.dispatch( CORE_SITE )
					.receiveGetFirstPartyModeSettings( {
						isEnabled: true,
					} );
			};

			return (
				<WithRegistrySetup func={ setupRegistry }>
					<Story />
				</WithRegistrySetup>
			);
		},
	],
};
