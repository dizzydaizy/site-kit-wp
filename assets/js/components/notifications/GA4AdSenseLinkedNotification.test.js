/**
 * GA4AdSenseLinkedNotification component tests.
 *
 * Site Kit by Google, Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * Internal dependencies
 */
import {
	render,
	createTestRegistry,
	provideModules,
	provideNotifications,
	act,
	waitForDefaultTimeouts,
} from '../../../../tests/js/test-utils';
import { provideAnalytics4MockReport } from '../../modules/analytics-4/utils/data-mock';
import { CORE_USER } from '../../googlesitekit/datastore/user/constants';
import { MODULES_ANALYTICS_4 } from '../../modules/analytics-4/datastore/constants';
import {
	VIEW_CONTEXT_MAIN_DASHBOARD,
	VIEW_CONTEXT_SETTINGS,
} from '../../googlesitekit/constants';
import { DEFAULT_NOTIFICATIONS } from '../../googlesitekit/notifications/register-defaults';
import { NOTIFICATION_AREAS } from '../../googlesitekit/notifications/datastore/constants';
import Notifications from './Notifications';

const GA4_ADSENSE_LINKED_NOTIFICATION =
	'top-earning-pages-success-notification';

describe( 'GA4AdSenseLinkedNotification', () => {
	let registry;

	const analyticsReport = new RegExp(
		'^/google-site-kit/v1/modules/analytics-4/data/report'
	);

	beforeEach( () => {
		registry = createTestRegistry();
		// All the below conditions will trigger a successful notification.
		// So each individual failing test case further below will overwrite one
		// of the success criteria.
		act( () => {
			provideModules( registry, [
				{
					active: true,
					connected: true,
					slug: 'analytics-4',
				},
				{
					active: true,
					connected: true,
					slug: 'adsense',
				},
			] );
			provideNotifications(
				registry,
				{
					[ GA4_ADSENSE_LINKED_NOTIFICATION ]:
						DEFAULT_NOTIFICATIONS[
							GA4_ADSENSE_LINKED_NOTIFICATION
						],
				},
				true
			);
			registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
				adSenseLinked: true,
			} );
			registry.dispatch( CORE_USER ).receiveGetDismissedItems( [] );
			fetchMock.getOnce( analyticsReport, {
				body: {
					rowCount: null,
				},
				status: 200,
			} );
		} );
	} );

	it( 'does not render if AdSense module is not active', async () => {
		act( () => {
			provideModules( registry, [
				{
					active: false,
					connected: false,
					slug: 'adsense',
				},
			] );
		} );

		const { container, waitForRegistry } = render(
			<Notifications areaSlug={ NOTIFICATION_AREAS.BANNERS_BELOW_NAV } />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await act( waitForRegistry );
		await act( waitForDefaultTimeouts );

		expect( container.childElementCount ).toBe( 0 );
	} );

	it( 'does not render if AdSense and Analytics are not linked', async () => {
		act( () => {
			registry.dispatch( MODULES_ANALYTICS_4 ).setSettings( {
				adSenseLinked: false,
			} );
		} );

		const { container, waitForRegistry } = render(
			<Notifications areaSlug={ NOTIFICATION_AREAS.BANNERS_BELOW_NAV } />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await act( waitForRegistry );
		await act( waitForDefaultTimeouts );

		expect( container.childElementCount ).toBe( 0 );
	} );

	it( 'does not render if report has data, and it is not already dismissed', async () => {
		const reportOptions = {
			startDate: '2020-08-11',
			endDate: '2020-09-07',
			dimensions: [ 'pagePath' ],
			metrics: [ { name: 'totalAdRevenue' } ],
			orderby: [
				{
					metric: { metricName: 'totalAdRevenue' },
					desc: true,
				},
			],
			limit: 3,
		};
		registry.dispatch( CORE_USER ).setReferenceDate( '2020-09-08' );
		provideAnalytics4MockReport( registry, reportOptions );

		const { container, waitForRegistry } = render(
			<Notifications areaSlug={ NOTIFICATION_AREAS.BANNERS_BELOW_NAV } />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await act( waitForRegistry );
		await act( waitForDefaultTimeouts );

		expect( fetchMock ).not.toHaveFetched(
			'/google-site-kit/v1/modules/analytics-4/data'
		);
		expect( container.childElementCount ).toBe( 0 );
	} );

	it( 'does not render if already dismissed', async () => {
		registry
			.dispatch( CORE_USER )
			.receiveGetDismissedItems( [ GA4_ADSENSE_LINKED_NOTIFICATION ] );

		const { container, waitForRegistry } = render(
			<Notifications areaSlug={ NOTIFICATION_AREAS.BANNERS_BELOW_NAV } />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await act( waitForRegistry );
		await act( waitForDefaultTimeouts );

		expect( fetchMock ).not.toHaveFetched(
			'/google-site-kit/v1/modules/analytics-4/data'
		);

		expect( container.childElementCount ).toBe( 0 );
	} );

	it( 'does not render when not on the main or entity dashboard', async () => {
		const { container, waitForRegistry } = render(
			<Notifications areaSlug={ NOTIFICATION_AREAS.BANNERS_BELOW_NAV } />,
			{
				registry,
				viewContext: VIEW_CONTEXT_SETTINGS,
			}
		);

		await act( waitForRegistry );
		await act( waitForDefaultTimeouts );

		expect( fetchMock ).not.toHaveFetched(
			'/google-site-kit/v1/modules/analytics-4/data'
		);

		expect( container.childElementCount ).toBe( 0 );
	} );

	it( 'renders when both Analytics & AdSense modules are active & linked, when report has no data and when it was not previously dismissed', async () => {
		const { container, waitForRegistry } = render(
			<Notifications areaSlug={ NOTIFICATION_AREAS.BANNERS_BELOW_NAV } />,
			{
				registry,
				viewContext: VIEW_CONTEXT_MAIN_DASHBOARD,
			}
		);

		await act( waitForRegistry );
		await act( waitForDefaultTimeouts );

		expect( container ).toHaveTextContent(
			'Your AdSense and Analytics accounts are linked'
		);
		expect( container ).toMatchSnapshot();
	} );
} );
