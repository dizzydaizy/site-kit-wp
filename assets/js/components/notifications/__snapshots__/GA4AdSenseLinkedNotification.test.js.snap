// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GA4AdSenseLinkedNotification renders when both Analytics & AdSense modules are active & linked, when report has no data and when it was not previously dismissed 1`] = `
<div>
  <section
    id="top-earning-pages-success-notification"
  >
    <div
      class="mdc-layout-grid"
    >
      <div
        class="mdc-layout-grid__inner"
      >
        <div
          class="googlesitekit-subtle-notification googlesitekit-subtle-notification--success mdc-layout-grid__cell mdc-layout-grid__cell--align-middle mdc-layout-grid__cell--span-12"
        >
          <div
            class="googlesitekit-subtle-notification__icon"
          >
            <svg />
          </div>
          <div
            class="googlesitekit-subtle-notification__content"
          >
            <p>
              Your AdSense and Analytics accounts are linked
            </p>
            <p
              class="googlesitekit-subtle-notification__secondary_description"
            >
              We’ll let you know as soon as there’s enough data available
            </p>
          </div>
          <div
            class="googlesitekit-subtle-notification__action"
          >
            <button
              class="mdc-button mdc-button--tertiary"
              target="_self"
            >
              <span
                class="mdc-button__label"
              >
                Got it
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
